##################################################
## R script for MetaboAnalyst
## Description: report generation using Sweave
## Note: most analyses were already performed, only need to embedding
## the results to the right place without rerun the whole analysis
## through Sweave. Only some auxilliary info (i.e. time, version etc need to
## run R throught Sweave
##
## Author: Jeff Xia, jeff.xia@mcgill.ca
## McGill University, Canada
##
## License: GNU GPL (>= 2)
###################################################


CreatePDFReport<-function(usrName){

    # create the Rnw file
    file.create("Analysis_Report.Rnw");
    # open for write
    rnwFile <<- file("Analysis_Report.Rnw", "w")

    # create a global counter to label figures
    fig.count <<- 0;
    table.count <<- 0;

    if(analSet$type == "stat" ){
        CreateStatRnwReport(usrName);
    }else if(analSet$type == "ts"){
        CreateTimeSeriesRnwReport(usrName);
    }else if(substr(analSet$type, 0, 4) == "mset"){
        CreateEnrichRnwReport(usrName);
    }else{
        CreatePathRnwReport(usrName);
    }
    # close opened files
    close(rnwFile);

    Sweave("Analysis_Report.Rnw");
    tools::texi2dvi("Analysis_Report.tex", pdf = TRUE, quiet=TRUE);
}

################################
## MetaboAnalyst statistics
###############################

# write .Rnw file template
CreateStatRnwReport<-function(usrName){

    CreateHeader(usrName);
    CreateStatIntr();

    CreateStatIOdoc();
    CreateStatNORMdoc();

    InitStatAnalMode();
    if(exists("analSet")){
        CreateUNIVdoc();
        CreateANOVAdoc();
        CreateCorrDoc();
        CreatePCAdoc();
        CreatePLSdoc();
        CreateSAMdoc();
        CreateEBAMdoc();
        CreateHCdoc();
        CreateKMdoc();
        CreateSOMdoc();
        CreateRFdoc();
        CreateSVMdoc();
    }else{
        CreateAnalNullMsg();
    }
    CreateStatFooter();
}

# create header
CreateHeader<-function(usrName){
        header <- c("\\documentclass[a4paper]{article}",
                "\\usepackage[margin=1.0in]{geometry}",
                "\\usepackage{longtable}",
                "\\SweaveOpts{eps=FALSE,pdf=TRUE}",
                "\\title{Metabolomic Data Analysis with MetaboAnalyst 3.0}",
                paste("\\author{ User ID: ", usrName, " }", sep=""),
                "\\begin{document}",
                "\\parskip=.3cm",
                "\\maketitle");
        cat(header, file=rnwFile, sep="\n", append=TRUE);
}

CreateStatIntr<-function(){
    descr <- c("\\section{Data Processing and Normalization}\n");
    cat(descr, file=rnwFile, append=TRUE);
}

# read and process the raw data
CreateStatIOdoc<-function(){
        descr <- c("\\subsection{Reading and Processing the Raw Data}\n",
                "MetaboAnalyst accepts a variety of data types generated in metabolomic studies,",
                "including compound concentration data, binned NMR/MS spectra data, NMR/MS peak",
                "list data, as well as MS spectra (NetCDF, mzXML, mzDATA).",
                "Users need to specify the data types when uploading their data in order for",
                "MetaboAnalyst to select the correct algorithm to process them.\n",
                paste("Table", table.count<<-table.count+1,"summarizes the result of the data processing steps.\n")
                );
        cat(descr, file=rnwFile, append=TRUE);

        # error checking
        if(is.null(dataSet$orig) | is.null(dataSet$proc) | is.null(dataSet$cls)){
                errorMsg<- c(descr, "Error occured during reading the raw data ....",
                    "Failed to proceed. Please check if the data format you uploaded is correct.",
                    "Please visit our FAQs, Data Formats, and TroubleShooting pages for more information!\n");
                cat(errorMsg, file=rnwFile, append=TRUE);
                return();
        }

        if(dataSet$type=="conc"){
            descr<-c("\\subsubsection{Reading Concentration Data}\n",
                "The concentration data should be uploaded in comma separated values (.csv) format.",
                "Samples can be in rows or columns, with class labels immediately following the sample IDs.\n");
            cat(descr, file=rnwFile, append=TRUE);
            cat("\n\n", file=rnwFile, append=TRUE);
            cat(dataSet$read.msg, file=rnwFile, append=TRUE, sep="\n");

         }else if(dataSet$type=="specbin"){
            descr<-c("\\subsubsection{Reading Binned Spectral Data}\n",
                "The binned spectra data should be uploaded in comma seperated values (.csv) format.",
                "Samples can be in rows or columns, with class labels immediately following the sample IDs.\n");
            cat(descr, file=rnwFile, append=TRUE);
            cat("\n\n", file=rnwFile, append=TRUE);
            cat(dataSet$read.msg, file=rnwFile, append=TRUE, sep="\n");

        }else if(dataSet$type=="pktable"){
            descr<-c("\\subsubsection{Reading Peak Intensity Table}\n",
                "The peak intensity table should be uploaded in comma separated values (.csv) format.",
                "Samples can be in rows or columns, with class labels immediately following the sample IDs.\n");
           cat(descr, file=rnwFile, append=TRUE);
           cat("\n\n", file=rnwFile, append=TRUE);
           cat(dataSet$read.msg, file=rnwFile, append=TRUE, sep="\n");
        }else if(dataSet$type=="nmrpeak"){
            descr<-c("\\subsubsection{Reading NMR Peak List and Intensities Data}\n",
                "NMR peak list and intensities data should be uploaded as one zip file. It contains subfolders, one for",
                "each group. Each folder contains peak list files, one per spectrum. The peak list format",
                "is a a two-column comma separated values - the first column indicates peak position (ppm)",
                "and the second one for peak intensities. The first line is assumed to be column labels.",
                "The files should be saved in .csv format. For paired analysis, users need to upload",
                "a text file specifying the paired information. Each pair is indicated by their sample names",
                "seperated by a colon \":\" with one pair per line.\n");
           cat(descr, file=rnwFile, append=TRUE);
           cat("\n\n", file=rnwFile, append=TRUE);
           cat(dataSet$read.msg, file=rnwFile, append=TRUE, sep="\n");

           descr<-c("\\subsubsection{Peak List Alignment}\n",
                "Proximal peaks are first grouped together based on their position using a moving window of 0.03 ppm",
                "and a step of 0.015 ppm. Peaks of the same group are aligned to their median positions across all samples.",
                "If more than one peak from the same sample appear in the same group, they will be replaced by their sum.",
                "Some peaks that are detected in very few samples (less than half in both classes) are excluded.",
                "The aligned peaks are reorganized into a single data matrix for further analysis. The name of the parent",
                "folder is used as class label for each sample.\n");
            cat(descr, file=rnwFile, append=TRUE);
            cat("\n\n", file=rnwFile, append=TRUE);
            cat(dataSet$proc.msg, file=rnwFile, append=TRUE, sep="\n");
         }else if(dataSet$type=="mspeak"){
            descr<-c("\\subsubsection{Reading MS Peak List and Intensities Data}\n",
                "MS peak list and intensities data should be uploaded as one zip file. It contains subfoulders with one for",
                "each group. Each folder contains peak list files, one per spectrum. The MS peak list format",
                "is either a two-column (mass and intensities) or three-column (mass, retention time, and intensities)",
                "comma separated values. The first line is assumed to be column labels.",
                "The files should be saved in .csv format. For paired analysis, users need to upload separately",
                "a text file specifying the paired information. Each pair is indicated by their sample names",
                "seperated by a colon \":\" with one pair per line.\n");
           cat(descr, file=rnwFile, append=TRUE);
           cat("\n\n", file=rnwFile, append=TRUE);
           cat(dataSet$read.msg, file=rnwFile, append=TRUE, sep="\n");

           descr<-c("\\subsubsection{Peak Matching and Alignment}\n",
                "Peaks need to be matched across samples in order to be compared. For two-column data, the",
                "program matches peaks by their m/z values. For three-column data, the program will further",
                "group peaks based on their retention time. During the process, mz and rt of each peak will",
                "be changed to their group median values. If a sample has more than one peak in a group,",
                "they will be replaced by their sum. Some peaks are excluded if they appear in less than half",
                "of both classes. The aligned peaks are reorganized into a single data matrix for further analysis.",
                "The name of the parent folder is used as class label for each sample.\n");
            cat(descr, file=rnwFile, append=TRUE);
            cat("\n\n", file=rnwFile, append=TRUE);
            cat(dataSet$proc.msg, file=rnwFile, append=TRUE, sep="\n");

         }else{ # spectra
            descr<-c("\\subsubsection{Reading GC/LC-MS Spectra}\n",
                "The spectra processing is carried out using the XCMS package.",
                "\\footnote{Colin A. Smith and Ralf Tautenhahn. \\textit{xcms: LC\\slash MS and",
                "GC\\slash MS Data Analysis}, 2008, R package version 1.14.0}",
                "Raw GC/LC-MS spectra can be in either NetCDF, mzXML or mzData format.",
                "You should create a separate folder for each group and create a",
                "single zip file to upload to MetaboAnalyst. The size limit for each uploaded file is 50M.\n");
           cat(descr, file=rnwFile, append=TRUE);
           cat("\n\n", file=rnwFile, append=TRUE);
           cat(dataSet$read.msg, file=rnwFile, append=TRUE, sep="\n");

           descr<-c("\\subsubsection{Peak Alignment and Retention Time Correction}\n",
                "In this step, the program automatically performs peak detection and peak alignment based on mass and",
                "retention time. The aligned peaks are reorganized into a data matrix for further analysis.",
                "Please note, the name of the parent folder is used as class label for each sample.\n");
           cat(descr, file=rnwFile, append=TRUE);
           cat("\n\n", file=rnwFile, append=TRUE);
           cat(c(dataSet$xset.msg, paste("Please see Figure", fig.count<<-fig.count+1, "for a summary graph.")), file=rnwFile, append=TRUE, sep="\n");
           cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$msrt,"}", sep=""),
                    "\\caption{Retention time deviation profiles used for sample alignment.
                        The data points used for generating each profile are also shown.
                        All times are in seconds. A negative number indicates a sample was eluting
                        before most of the others, and vice versa.}",
                    "\\end{center}",
                    paste("\\label{",imgSet$msrt,"}", sep=""),
                    "\\end{figure}"
                );
            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
         }

        # the last step is sanity check
        descr<-c("\\subsubsection{Data Integrity Check}\n",
                "Before data analysis, a data integrity check is performed to make sure that all the necessary",
                "information has been collected. The class labels must be present and contain only two classes.",
                "If samples are paired, the class label must be from -n/2 to -1 for one group, and 1 to n/2 for the other group",
                "(n is the sample number and must be an even number). Class labels with same absolute value are assumed to be pairs.",
                "Compound concentration or peak intensity values should all be non-negative numbers.",
                "By default, all missing values, zeros and negative values will be replaced by the half of the minimum positive value",
                "found within the data (see next section)");
        cat(descr, file=rnwFile, append=TRUE);
        cat("\n\n", file=rnwFile, append=TRUE);

        descr<-c("\\subsubsection{Missing value imputations}\n",
                "Too many zeroes or missing values will cause difficulties for downstream analysis.",
                "MetaboAnalyst offers several different methods for this purpose. The default method replaces ",
                "all the missing and zero values with a small values (the half of the minimum positive",
                "values in the original data) assuming to be the detection limit. The assumption of this approach",
                "is that most missing values are caused by low abundance metabolites (i.e.below the detection limit).",
                "In addition, since zero values may cause problem for data normalization (i.e. log), they are also ",
                "replaced with this small value. User can also specify other methods, such as replace by mean/median,",
                "or use K-Nearest Neighbours, Probabilistic PCA (PPCA), Bayesian PCA (BPCA) method, Singular Value Decomposition (SVD)",
                "method to impute the missing values \\footnote{Stacklies W, Redestig H, Scholz M, Walther D, Selbig J.",
                "\\textit{pcaMethods: a bioconductor package, providing PCA methods for incomplete data.}, Bioinformatics",
                "2007 23(9):1164-1167}. Please choose the one that is the most appropriate for your data.");
        cat(descr, file=rnwFile, append=TRUE);
        cat("\n\n", file=rnwFile, append=TRUE);

        cat(dataSet$replace.msg, file=rnwFile, append=TRUE);
        cat("\n\n", file=rnwFile, append=TRUE);


        # the data filtering
        descr<-c("\\subsubsection{Data Filtering}\n",
                "The purpose of the data filtering is to identify and remove variables that are unlikely to be of",
                "use when modeling the data. No phenotype information are used in the filtering process, so the result",
                "can be used with any downstream analysis. This step can usually improves the results.",
                "Data filter is strongly recommended for datasets with large number of variables (> 250)",
                "datasets contain much noise (i.e.chemometrics data). Filtering can usually improve your",

                 "results\\footnote{Hackstadt AJ, Hess AM.\\textit{Filtering for increased power for microarray data analysis},",
                 "BMC Bioinformatics. 2009; 10: 11.}.",
                 "\n\n",
                "\\textit{For data with number of variables  < 250, this step will reduce 5\\% of variables;",
                "For variable number between 250 and 500, 10\\% of variables will be removed;",
                "For variable number bwteen 500 and 1000, 25\\% of variables will be removed;",
                "And 40\\% of variabled will be removed for data with over 1000 varaibles.}");
        cat(descr, file=rnwFile, append=TRUE);
        cat("\n\n", file=rnwFile, append=TRUE);

        filt.msg <- dataSet$filter.msg;
        if(is.null(filt.msg)){
            filt.msg <- "No data filtering was performed.";
        }
        cat(filt.msg, file=rnwFile, append=TRUE);
        cat("\n\n", file=rnwFile, append=TRUE);

        cmdhist2<-c("<<echo=false, results=tex>>=",
                    "CreateSummaryTable()",
                    "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");
}

# create normalization doc 
CreateStatNORMdoc<-function(){

        # need to check if this process is executed
        if(is.null(dataSet$norm)){
                errorMsg<- c("Error occured during normalization of your data ....",
                    "Fail to proceed. Please check if the data format you uploaded is correct.",
                    "Please visit our FAQs, Data Formats, and TroubleShooting pages for more information!");
                cat(errorMsg, file=rnwFile, append=TRUE);
                return();
        }

        descr1 <- c("\\subsection{Data Normalization}\n",
                "The data is stored as a table with one sample per row and one variable (bin\\slash peak\\slash",
                "metabolite) per column. The normalization procedures implemented below are grouped into four categories.",
                "Sample specific normalization allows users to manually adjust concentrations based on biological inputs",
                "(i.e. volume, mass); row-wise normalization allows general-purpose adjustment for differences among samples;",
                "data transformation and scaling are two different approaches to make features more comparable.",
                "You can use one or combine both to achieve better results.",
                "\n\n",
                "The normalization consists of the following options:");

        cat(descr1, file=rnwFile, append=TRUE);
        descr2 <- c("\\begin{enumerate}",
                "\\item{Sample specific normalization (i.e. normalize by dry weight, volume) }",
                "\\item{Row-wise procedures: }",
                "\\begin{itemize}",
                "\\item{Normalization by the sum }",
                "\\item{Normalization by the sample median }",
                "\\item{Normalization by a reference sample (probabilistic quotient",
                    "normalization)\\footnote{Dieterle F, Ross A, Schlotterbeck G, Senn H. \\textit{Probabilistic quotient normalization as robust",
                    "method to account for dilution of complex biological mixtures. Application in 1H NMR metabonomics}, 2006,",
                    "Anal Chem 78 (13);4281 - 4290}}",
                "\\item{Normalization by a reference feature (i.e. creatinine, internal control) }",
                "\\end{itemize}",
                "\\item{Data transformation : }",
                "\\begin{itemize}",
                "\\item{Generalized log transformation (glog 2) }",
                "\\item{Cube root transformation}",
                "\\end{itemize}",
                "\\item{Data scaling: }",
                "\\begin{itemize}",
                "\\item{Unit scaling (mean-centered and divided by standard deviation of each variable)}",
                "\\item{Pareto scaling (mean-centered and divided by the square root of standard deviation of each variable)}",
                "\\item{Range scaling (mean-centered and divided by the value range of each variable)}",
                "\\end{itemize}",
                "\\end{enumerate}",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the effects before and after normalization.\n"));

        cat(descr2, file=rnwFile, append=TRUE, sep="\n");

        if(dataSet$combined.method){
            norm.desc <- "Combined approach using quantile normalization within replicates after log transformation.";
        }else{
            norm.desc <- paste("Row-wise normalization: ", dataSet$rownorm.method, "; ",
                    "Data transformation: ",dataSet$trans.method, "; ",
                    "Data scaling: ",dataSet$scale.method, ".", sep="");
        }
        cmdhist<-c( "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$norm,"}", sep=""),
                    "\\caption{Box plots and kernel density plots before and after normalization.",
                    "The boxplots show at most 50 features due to space limit. The density plots are based on all samples.",
                    "Selected methods :", norm.desc, "}",
                    "\\end{center}",
                    paste("\\label{",imgSet$norm,"}", sep=""),
                    "\\end{figure}",
                    "\\clearpage"
                );
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
}

InitStatAnalMode<-function(){
        descr <- c("\\section{Statistical and Machine Learning Data Analysis}",
                   "MetaboAnalyst offers a variety of methods commonly used in metabolomic data analyses.",
                   "They include:\n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");

        descr2 <- c(
                "\\begin{enumerate}",
                "\\item{Univariate analysis methods: }",
                "\\begin{itemize}",
                "\\item{Fold Change Analysis }",
                "\\item{T-tests}",
                "\\item{Volcano Plot}",
                "\\item{One-way ANOVA and post-hoc analysis}",
                "\\item{Correlation analysis}",
                "\\end{itemize}",
                "\\item{Multivariate analysis methods: }",
                "\\begin{itemize}",
                "\\item{Principal Component Analysis (PCA) }",
                "\\item{Partial Least Squares - Discriminant Analysis (PLS-DA) }",
                "\\end{itemize}",
                "\\item{Robust Feature Selection Methods in microarray studies }",
                "\\begin{itemize}",
                "\\item{Significance Analysis of Microarray (SAM)}",
                "\\item{Empirical Bayesian Analysis of Microarray (EBAM)}",
                "\\end{itemize}",
                "\\item{Clustering Analysis}",
                "\\begin{itemize}",
                "\\item{Hierarchical Clustering}",
                "\\begin{itemize}",
                "\\item{Dendrogram}",
                "\\item{Heatmap}",
                "\\end{itemize}",
                "\\item{Partitional Clustering}",
                "\\begin{itemize}",
                "\\item{K-means Clustering}",
                "\\item{Self-Organizing Map (SOM)}",
                "\\end{itemize}",
                "\\end{itemize}",
                "\\item{Supervised Classification and Feature Selection methods}",
                "\\begin{itemize}",
                "\\item{Random Forest}",
                "\\item{Support Vector Machine (SVM)}",
                "\\end{itemize}",
                "\\end{enumerate}",
                "\\texttt{Please note: some advanced methods are available only for two-group sample analyais.}",
                "\\clearpage"
            );
        cat(descr2, file=rnwFile, append=TRUE, sep="\n");
}

CreateAnalNullMsg<-function(){
        descr <- c("No analysis was performed on your data.\n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

# create Univartests doc
CreateUNIVdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$fc) & is.null(analSet$tt) & is.null(analSet$volcano)){
                return();
        }

        if(isEmptyMatrix(analSet$fc$sig.mat)){
            fc.tab<-NULL;
        }else{
            fc.tab<-paste("Table", table.count<<-table.count+1,"shows the details of these features;");
        }
        if(isEmptyMatrix(analSet$tt$sig.mat)){
            tt.tab<-NULL;
        }else{
            tt.tab<-paste("Table", table.count<<-table.count+1,"shows the details of these features;");
        }
        if(isEmptyMatrix(analSet$volcano$sig.mat)){
            volcano.tab<-NULL;
        }else{
            volcano.tab<-paste("Table", table.count<<-table.count+1,"shows the details of these features.");
        }


        descr <- c("\\subsection{Univariate Analysis}\n",
                "Univariate analysis methods are the most common methods used for exploratory data analysis. ",
                "For two-group data, MetaboAnalyst provides Fold Change (FC) analysis, t-tests, and volcano",
                "plot which is a combination of the first two methods. All three these methods support both",
                "unpaired and paired analyses. For multi-group analysis, MetaboAnalyst provides two types of",
                "analysis - one-way analysis of variance (ANOVA) with associated post-hoc analyses, and correlation",
                "analysis to identify signficant compounds that follow a given pattern. The univariate analyses provide",
                "a preliminary overview about features that are potentially significant in discriminating",
                "the conditions under study.",
                "\n\n",
                "For paired fold change analysis, the algorithm first counts the total number of pairs with fold changes",
                "that are consistently above/below the specified FC threshold for each variable. A variable will be",
                "reported as significant if this number is above a given count threshold (default > 75\\% of pairs/variable)",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the important features identified by fold change analysis."),
                fc.tab,
                paste("Figure", fig.count<<-fig.count+1,"shows the important features identified by t-tests."),
                tt.tab,
                paste("Figure", fig.count<<-fig.count+1,"shows the important features identified by volcano plot."),
                volcano.tab,
                "\n\n",
                "Please note, the purpose of fold change is to compare absolute value changes between two group means.",
                "Therefore, the data before column normlaization will be used instead. Also note, the result is plotted",
                "in log2 scale, so that same fold change (up/down-regulated) will have the same distance to the zero baseline.",
                "\n");

        cat(descr, file=rnwFile, append=TRUE);

        # Fold change
        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$fc,"}", sep=""),
                    "\\caption{", paste("Important features selected by fold-change analysis with threshold ", analSet$fc$raw.thresh, ". ",
                       "The red circles represent features above the threshold. Note the values are on log scale, so that both up-regulated ",
                       "and downregulated features can be plotted in a symmetrical way", sep=""), "}",
                    "\\end{center}",
                    paste("\\label{",imgSet$fc,"}", sep=""),
                    "\\end{figure}"
                );
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\n\n", file=rnwFile, append=TRUE, sep="\n");
        cmdhist2<-c("<<echo=false, results=tex>>=",
                    "GetSigTable.FC()",
                    "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");

        # T-tests
        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$tt,"}", sep=""),
                    "\\caption{", paste("Important features selected by t-tests with threshold ", analSet$tt$raw.thresh, ". ",
                    "The red circles represent features above the threshold. Note the p values are transformed by -log10 so that the more significant ",
                    "features (with smaller p values) will be plotted higher on the graph. ", sep=""),"}",
                    "\\end{center}",
                    paste("\\label{",imgSet$tt,"}", sep=""),
                    "\\end{figure}"
                );
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\n\n", file=rnwFile, append=TRUE, sep="\n");
        cmdhist2<-c("<<echo=false, results=tex>>=",
                    "GetSigTable.TT()",
                    "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");

        # Volcano plot
        cmdhist<-c( 
                  "\\begin{figure}[htp]",
                  "\\begin{center}",
                  paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$volcano,"}", sep=""),
                  "\\caption{", paste("Important features selected by volcano plot with fold change threshold (x) ",
                    analSet$volcano$raw.threshx, " and t-tests threshold (y) ", analSet$volcano$raw.threshy, ". ",
                    "The red circles represent features above the threshold. Note both fold changes and p values are log ",
                    "transformed. The further its position away from the (0,0), the more significant the feature is. ", sep=""),"}",
                  "\\end{center}",
                  paste("\\label{",imgSet$volcano,"}", sep=""),
                  "\\end{figure}"
              );
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\n\n", file=rnwFile, append=TRUE, sep="\n");
        cmdhist2<-c("<<echo=false, results=tex>>=",
                   "GetSigTable.Volcano()",
                   "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");

        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");
}

# create Univartests doc
CreateANOVAdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$aov)){
                return();
        }

        if(isEmptyMatrix(analSet$aov$sig.mat)){
            anova.tab<-NULL;
        }else{
            anova.tab<-paste("Table", table.count<<-table.count+1,"shows the details of these features.",
                "The \\texttt{post-hoc Sig. Comparison} column shows the comparisons between different levels",
                "that are significant given the p value threshold. ");
        }

        descr <- c("\\subsection{One-way ANOVA}\n",
                "Univariate analysis methods are the most common methods used for exploratory data analysis. ",
                "For multi-group analysis, MetaboAnalyst provides one-way Analysis",
                "of Variance (ANOVA). As ANOVA only tells whether the overall comparison is significant or not,",
                "it is usually followed by post-hoc analyses in order to identify which two levels are different.",
                "MetaboAnalyst provides two most commonly used methods for this purpose - Fisher's",
                "least significant difference method (Fisher's LSD) and Tukey's Honestly Significant Difference",
                "(Tukey's HSD). The univariate analyses provide a preliminary overview about features that are",
                "potentially significant in discriminating the conditions under study.",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the important features identified by ANOVA analysis."),
                anova.tab,
                "\n");

        cat(descr, file=rnwFile, append=TRUE);

        # ANOVA
        cmdhist<-c(
                  "\\begin{figure}[htp]",
                  "\\begin{center}",
                  paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$anova,"}", sep=""),
                  "\\caption{", paste("Important features selected by ANOVA plot with p value threshold ",
                    analSet$aov$raw.thresh, ". ", sep=""),"}",
                  "\\end{center}",
                  paste("\\label{",imgSet$anova,"}", sep=""),
                  "\\end{figure}"
              );
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\n\n", file=rnwFile, append=TRUE, sep="\n");

        cmdhist2<-c("<<echo=false, results=tex>>=",
                   "GetSigTable.Anova()",
                   "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");

        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");
}

# create Univartests doc
CreateCorrDoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$cor.res) & is.null(imgSet$corr.heatmap)){
                return();
        }

        # need to check if this process is executed
        if(!is.null(imgSet$corr.heatmap)){
            descr <- c("\\subsection{Correlation Analysis}\n",
                "Correlation analysis can be used to visualize the overall correlations between different features",
                "It can also be used to identify which features are correlated with a feature of interest.",
                "Correlation analysis can also be used to identify if certain features show particular patterns",
                "under different conditions. Users first need to define a pattern in the form of a series of hyphenated numbers.",
                "For example, in a time-series study with four time points, a pattern of of",
                "\\texttt{1-2-3-4} is used to search compounds with increasing the concentration as",
                "time changes; while a pattern of \\texttt{3-2-1-3} can be used to search compounds",
                "that decrease at first, then bounce back to the original level.",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1, "shows the overall correlation heatmap."),
                "\n");

            cat(descr, file=rnwFile, append=TRUE);

            cmdhist<-c(
                  "\\begin{figure}[htp]",
                  "\\begin{center}",
                  paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$corr.heatmap,"}", sep=""),
                  "\\caption{Correlation Heatmaps}",
                  "\\end{center}",
                  paste("\\label{",imgSet$corr.heatmap,"}", sep=""),
                  "\\end{figure}"
              );

            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
            cat("\n\n", file=rnwFile, append=TRUE, sep="\n");
        }

        if(!is.null(analSet$cor.res)){
            if(isEmptyMatrix(analSet$cor.res)){
                cor.tab<-NULL;
            }else{
                cor.tab<-paste("Table", table.count<<-table.count+1,"shows the details of these features.");
            }

            descr <- c("\\subsection{Correlation Analysis}\n",
                "Correlation analysis can be used to identify which features are correlated with a feature of interest.",
                "Correlation analysis can also be used to identify if certain features show particular patterns",
                "under different conditions. Users first need to define a pattern in the form of a series of hyphenated numbers.",
                "For example, in a time-series study with four time points, a pattern of of",
                "\\texttt{1-2-3-4} is used to search compounds with increasing the concentration as",
                "time changes; while a pattern of \\texttt{3-2-1-3} can be used to search compounds",
                "that decrease at first, then bounce back to the original level.",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1, "shows the important features identified by correlation analysis."),
                cor.tab,
                "\n");

            cat(descr, file=rnwFile, append=TRUE);

            cmdhist<-c(
                  "\\begin{figure}[htp]",
                  "\\begin{center}",
                  paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$corr,"}", sep=""),
                  "\\caption{Important features selected by correlation analysis with light",
                  "purple indicates positive correlation and blue indicate negative correlations.}",
                  "\\end{center}",
                  paste("\\label{",imgSet$corr,"}", sep=""),
                  "\\end{figure}"
              );
            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
            cat("\n\n", file=rnwFile, append=TRUE, sep="\n");

            cmdhist2<-c("<<echo=false, results=tex>>=",
                   "GetSigTable.Corr()",
                   "@");
            cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        }

        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");
}

# create PCA doc
CreatePCAdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$pca)){
                return();
        }

        descr <- c("\\subsection{Principal Component Analysis (PCA)}\n",
                "PCA is an unsupervised method aiming to find the directions that best",
                "explain the variance in a data set (X) without referring to class labels (Y).",
                "The data are summarized into much fewer variables called \\textit{scores} which",
                "are weighted average of the original variables. The weighting profiles are called",
                "\\textit{loadings}. The PCA analysis is performed using the \\texttt{prcomp} package.",
                "The calculation is based on singular value decomposition.",
                "\n\n",
                "The Rscript \\texttt{chemometrics.R} is required.",
                paste("Figure", fig.count<<-fig.count+1,"is pairwise score plots providing an overview of the various seperation patterns among the most significant PCs;"),
                paste("Figure", fig.count<<-fig.count+1,"is the scree plot showing the variances explained by the selected PCs;"),
                paste("Figure", fig.count<<-fig.count+1,"shows the 2-D scores plot between selected PCs;"),
                paste("Figure", fig.count<<-fig.count+1,"shows the 3-D scores plot between selected PCs;"),
                paste("Figure", fig.count<<-fig.count+1,"shows the loadings plot between the selected PCs;"),
                paste("Figure", fig.count<<-fig.count+1,"shows the biplot between the selected PCs.\n"));

       cat(descr, file=rnwFile, append=TRUE);

       cmdhist<-c(
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pca.pair,"}", sep=""),
                "\\caption{Pairwise score plots between the selected PCs. The explained variance of each PC is shown in the corresponding diagonal cell. }",
                "\\end{center}",
                paste("\\label{",imgSet$pca.pair,"}", sep=""),
                "\\end{figure}",
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pca.scree,"}", sep=""),
                "\\caption{Scree plot shows the variance explained by PCs. The green line on top shows the accumulated variance explained; the blue line underneath shows the variance explained by individual PC.}",
                "\\end{center}",
                paste("\\label{",imgSet$pca.scree,"}", sep=""),
                "\\end{figure}",
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pca.score2d,"}", sep=""),
                "\\caption{Scores plot between the selected PCs. The explained variances are shown in brackets.}",
                "\\end{center}",
                paste("\\label{",imgSet$pca.score2d,"}", sep=""),
                "\\end{figure}",
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pca.score3d,"}", sep=""),
                "\\caption{3D score plot between the selected PCs. The explained variances are shown in brackets.}",
                "\\end{center}",
                paste("\\label{",imgSet$pca.score3d,"}", sep=""),
                "\\end{figure}",
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pca.loading,"}", sep=""),
                "\\caption{Loadings plot for the selected PCs. }",
                "\\end{center}",
                paste("\\label{",imgSet$pca.loading,"}", sep=""),
                "\\end{figure}",
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pca.biplot,"}", sep=""),
                "\\caption{PCA biplot between the selected PCs. Note, you may want to test different centering and scaling
                    normalization methods for the biplot to be displayed properly.}",
                "\\end{center}",
                paste("\\label{",imgSet$pca.biplot,"}", sep=""),
                "\\end{figure}"
               );
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");
}

# create PLS doc
CreatePLSdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$plsr) & is.null(analSet$plsda)){
                return();
        }

        descr <- c("\\subsection{Partial Least Squares - Discriminant Analysis (PLS-DA)}\n",
                "PLS is a supervised method that uses multivariate regression techniques to extract via",
                "linear combination of original variables (X) the information that can predict the",
                "class membership (Y). The PLS regression is performed using the \\texttt{plsr} function",
                "provided by R \\texttt{pls} package\\footnote{Ron Wehrens and Bjorn-Helge Mevik.\\textit{pls: Partial Least",
                "Squares Regression (PLSR) and Principal Component Regression (PCR)}, 2007,",
                "R package version 2.1-0}. The classification and cross-validation are performed using the corresponding wrapper",
                "function offered by the \\texttt{caret} package\\footnote{Max Kuhn. Contributions from",
                "Jed Wing and Steve Weston and Andre Williams.\\textit{caret: Classification and Regression",
                "Training}, 2008, R package version 3.45}.",
                "\n\n",
                "To assess the significance of class discrimination, a permutation test was performed. In each permutation, a PLS-DA model was",
                "built between the data (X) and the permuted class labels (Y) using the optimal number of components determined",
                "by cross validation for the model based on the original class assignment. MetaboAnalyst supports two types of test",
                "statistics for measuring the class discrimination. The first one is based on prediction accuracy during training.",
                "The second one is separation distance based on the ratio of the between group sum of the squares and the within",
                "group sum of squares (B/W-ratio).",
                "If the observed test statistic is part of the distribution based on the permuted class assignments,",
                "the class discrimination cannot be considered significant from a statistical point of",
                "view.\\footnote{Bijlsma et al.\\textit{Large-Scale Human Metabolomics Studies: A Strategy for Data",
                "(Pre-) Processing and Validation}, Anal Chem. 2006, 78 567 - 574}.",
                "\n\n",
                "There are two variable importance measures in PLS-DA. The first, Variable Importance in Projection (VIP) is",
                "a weighted sum of squares of the PLS loadings taking into account the amount of explained Y-variation",
                "in each dimension. Please note, VIP scores are calculated for each components. When more than componetnts are used to calculate", 
                "the feature importance, the average of the VIP scores are used. The other importance measure is based",
                "on the weighted sum of PLS-regression. The weights are a function of the reduction of the sums of squares across the number",
                "of PLS components. Please note, for multiple-group (more than two) analysis, the same number of predictors will be built for each",
                "group. Therefore, the coefficient of each feature will be different depending on which group you want to predict.",
                "The average of the feature coefficients are used to indicate the overall coefficient-based importance. ",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the overview of scores plots;"),
                paste("Figure", fig.count<<-fig.count+1,"shows the 2-D scores plot between selected components;"),
                paste("Figure", fig.count<<-fig.count+1,"shows the 3-D scores plot between selected components;"),
                paste("Figure", fig.count<<-fig.count+1,"shows the loading plot between the selected components;"));
        cat(descr, file=rnwFile, append=TRUE);

       
        descr<-c(paste("Figure", fig.count<<-fig.count+1,"shows the classification performance with different number of components;"),
                paste("Figure", fig.count<<-fig.count+1,"shows the results of permutation test for model validation;"),
                paste("Figure", fig.count<<-fig.count+1,"shows important features identified by PLS-DA.\n"));
        cat(descr, file=rnwFile, append=TRUE);


        plsrhist<-c(
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pls.pair,"}", sep=""),
                "\\caption{Pairwise scores plots between the selected components. The explained variance of each component is shown in the corresponding diagonal cell. }",
                "\\end{center}",
                paste("\\label{",imgSet$pls.pair,"}", sep=""),
                "\\end{figure}",
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pls.score2d,"}", sep=""),
                "\\caption{Scores plot between the selected PCs. The explained variances are shown in brackets. }",
                "\\end{center}",
                paste("\\label{",imgSet$pls.score2d,"}", sep=""),
                "\\end{figure}",
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pls.score3d,"}", sep=""),
                "\\caption{3D scores plot between the selected PCs. The explained variances are shown in brackets.}",
                "\\end{center}",
                paste("\\label{",imgSet$pls.score3d,"}", sep=""),
                "\\end{figure}",
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pls.loading,"}", sep=""),
                "\\caption{Loadings plot between the selected PCs. }",
                "\\end{center}",
                paste("\\label{",imgSet$pls.loading,"}", sep=""),
                "\\end{figure}"
            );
          cat(plsrhist, file=rnwFile, append=TRUE, sep="\n");

          plsdahist <- c(
                # classification fig
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pls.class,"}", sep=""),
                "\\caption{PLS-DA classification using different number of components. The red circle indicates the best classifier.}",
                "\\end{center}",
                paste("\\label{",imgSet$pls.class,"}", sep=""),
                "\\end{figure}"
            );
           cat(plsdahist, file=rnwFile, append=TRUE, sep="\n");
           
           plsdahist <- c(
                 # permutation fig
                 "\\begin{figure}[htp]",
                 "\\begin{center}",
                 paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pls.permut,"}", sep=""),
                 paste("\\caption{PLS-DA model validation by permutation tests based on ", analSet$plsda$permut.type, ". ",
                      "The p value based on permutation is ", analSet$plsda$permut.p, ".", sep=""), "}",
                 "\\end{center}",
                 paste("\\label{",imgSet$pls.permut,"}", sep=""),
                 "\\end{figure}"
            );
           cat(plsdahist, file=rnwFile, append=TRUE, sep="\n");
            

          plsdahist <- c(
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$pls.imp,"}", sep=""),
                paste("\\caption{Important features identified by PLS-DA. The colored boxes on the right indicate the relative concentrations of the corresponding metabolite in each group
                        under study. }"),
                "\\end{center}",
                paste("\\label{",imgSet$pls.imp,"}", sep=""),
                "\\end{figure}"
          );
         cat(plsdahist, file=rnwFile, append=TRUE, sep="\n");
          
         
         cat("\\clearpage", file=rnwFile, append=TRUE);
}

# create SAM doc
CreateSAMdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$sam)){
                return();
        }

        if(isEmptyMatrix(analSet$sam.cmpds)){
            sam.tab<-NULL;
        }else{
            sam.tab<-paste("Table", table.count<<-table.count+1,"shows the details of these features.");
        }

        descr <- c("\\subsection{Significance Analysis of Microarray (SAM)}\n",
                "SAM is a well-established statistical method for identification",
                "of differentially expressed genes in microarray data analysis. It is designed",
                "to address the false discovery rate (FDR) when running multiple tests on high-dimensional",
                "microarray data. SAM assigns a significance score to each variable based on its change",
                "relative to the standard deviation of repeated measurements. For a variable with scores",
                "greater than an adjustable threshold, its relative difference is compared to the",
                "distribution estimated by random permutations of the class labels. For each threshold,",
                "a certain proportion of the variables in the permutation set will be found to be significant",
                "by chance. The proportion is used to calculate the FDR. SAM is performed using the",
                "\\texttt{siggenes} package\\footnote{Holger Schwender. \\textit{siggenes: Multiple testing using",
                "SAM and Efron's empirical Bayes approaches},2008, R package version 1.16.0}.",
                "Users need to specify the \\texttt{Delta} value to control FDR in order to proceed.",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the significant features identified by SAM."),
                sam.tab,
                "\n");

        cat(descr, file=rnwFile, append=TRUE);
        
        cmdhist<-c( 
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$sam.cmpd,"}", sep=""),
                    "\\caption{Significant features identified by SAM. The green circles represent features
                        that exceed the specified threshold. }",
                    "\\end{center}",
                    paste("\\label{",imgSet$sam.cmpd,"}", sep=""),
                    "\\end{figure}"
                );
         cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
         cat("\n\n", file=rnwFile, append=TRUE, sep="\n");
         cmdhist2<-c("<<echo=false, results=tex>>=",
                    "GetSigTable.SAM()",
                    "@");
         cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
         cat("\\clearpage", file=rnwFile, append=TRUE);
}

# create EBAM doc, please note
# the search for delta (SAM) and a0 (EBAM) will not be plotted
# it is only exploration, and may cause potential inconsistentcies. 
CreateEBAMdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$ebam)){
                return();
        }

        if(isEmptyMatrix(analSet$ebam.cmpds)){
            ebam.tab<-NULL;
        }else{
            ebam.tab<-paste("Table", table.count<<-table.count+1,"shows the details of these features.");
        }

        descr <- c("\\subsection{Empirical Bayesian Analysis of Microarray (EBAM)}\n",
                "EBAM is an empirical Bayesian method based on moderated t-statistics.",
                "EBAM uses a two-group mixture model for null and significant features.",
                "The prior and density parameters are estimated from the data. A feature is",
                "considered significant if its calculated posterior is larger than or equal to",
                "\\texttt{delta} and no other features with a more extreme test score that",
                "is not called signicant. The default is \\texttt{delta} = 0.9.",
                "The suggested fudge factor (\\texttt{a0}) is chosen that leads to the largest number",
                "of significant features. EBAM is performed with \\texttt{ebam} function in",
                "\\texttt{siggenes} package\\footnote{Holger Schwender. \\textit{siggenes: Multiple testing using",
                "SAM and Efron's empirical Bayes approaches},2008,R package version 1.16.0}.",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the important features identified by EBAM."),
                ebam.tab,
                "\n");

        cat(descr, file=rnwFile, append=TRUE);
        
        cmdhist<-c( 
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$ebam.cmpd,"}", sep=""),
                    "\\caption{Significant features identified by EBAM. The green circles represent features
                        that exceed the specified threshold. }",
                    "\\end{center}",
                    paste("\\label{",imgSet$ebam.cmpd,"}", sep=""),
                    "\\end{figure}"
                    );

         cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
         cat("\n\n", file=rnwFile, append=TRUE, sep="\n");
         cmdhist2<-c("<<echo=false, results=tex>>=",
                    "GetSigTable.EBAM()",
                    "@");
         cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
         cat("\\clearpage", file=rnwFile, append=TRUE);
}

# hierachical clustering
CreateHCdoc<-function(){


        # need to check if this process is executed
        if(is.null(analSet$tree) & is.null(analSet$htmap)){
                return();
        }

        descr <- c("\\subsection{Hierarchical Clustering}\n",
                "In (agglomerative) hierarchical cluster analysis, each sample begins",
                " as a separate cluster and the algorithm proceeds to combine them until all",
                "samples belong to one cluster. Two parameters need to be considered when performing",
                "hierarchical clustering. The first one is similarity measure - Euclidean distance,",
                "Pearson's correlation, Spearman's rank correlation. The other parameter is clustering",
                "algorithms, including average linkage (clustering uses the centroids of the observations),",
                "complete linkage (clustering uses the farthest pair of observations between the two groups),",
                "single linkage (clustering uses the closest pair of observations) and Ward's linkage",
                "(clustering to minimize the sum of squares of any two clusters). Heatmap is often presented",
                "as a visual aid in addition to the dendrogram.",
                "\n\n",
                "Hierachical clustering is performed with the \\texttt{hclust} function in package \\texttt{stat}.",
                paste("Figure", fig.count<<-fig.count+1,"shows the clustering result in the form of a dendrogram."),
                paste("Figure", fig.count<<-fig.count+1,"shows the clustering result in the form of a heatmap.\n"));

        cat(descr, file=rnwFile, append=TRUE);

        if(!is.null(analSet$tree)){
            cmdhist<-c(
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$tree,"}", sep=""),
                paste("\\caption{Clustering result shown as dendrogram (", 
                      "distance measure using ", "\\texttt{", analSet$tree$dist.par, "}, and clustering algorithm using ", "\\texttt{", analSet$tree$clust.par, "}).}", sep=""),
                "\\end{center}",
                paste("\\label{",imgSet$tree,"}", sep=""),
                "\\end{figure}"
                );
            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        }
        if(!is.null(analSet$htmap)){
            cmdhist<-c(
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$heatmap,"}", sep=""),
                paste("\\caption{Clustering result shown as heatmap (", 
                      "distance measure using ", "\\texttt{", analSet$htmap$dist.par, "}, and clustering algorithm using ", "\\texttt{", analSet$htmap$clust.par, "}).}", sep=""),
                "\\end{center}",
                paste("\\label{",imgSet$heatmap,"}", sep=""),
                "\\end{figure}"
                );                    
            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        }
        cat("\\clearpage", file=rnwFile, append=TRUE);
}

# SOM partitional clustering
CreateSOMdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$som)){
                return();
        };

        descr <- c("\\subsection{Self Organizing Map (SOM)}\n",
                "SOM is an unsupervised neural network algorithm used to automatically",
                "identify major trends present in high-dimensional data. SOM is based",
                "on a grid of interconnected nodes, each of which represents a model.",
                "These models begin as random values, but during the process of iterative training",
                "they are updated to represent different subsets of the training set.",
                "Users need to specify the x and y dimension of the grid to perform SOM analysis.",
                "\n\n",
                "The SOM is performed using the R \\texttt{som} package\\footnote{Jun Yan. \\textit{som:",
                "Self-Organizing Map}, 2004, R package version 0.3-4}.",
                paste("Figure", fig.count<<-fig.count+1,"shows the SOM clustering results."),
                paste("Table", table.count<<-table.count+1,"shows the members in each cluster from SOM analysis.\n"));

        cat(descr, file=rnwFile, append=TRUE);

        cmdhist<-c( 
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$som,"}", sep=""),
                    "\\caption{SOM cluster analysis. The x-axes are features and y-axes are relative",
                    "intensities. The blue lines represent median intensities of corresponding clusters}",
                    "\\end{center}",
                    paste("\\label{",imgSet$som,"}", sep=""),
                    "\\end{figure}"
                    );
                    
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\n\n", file=rnwFile, append=TRUE, sep="\n");
        cmdhist2<-c("<<echo=false, results=tex>>=",
                    "GetAllSOMClusterMembers()",
                    "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE);
}

# Kmeans partitional clustering
CreateKMdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$kmeans)){
                return();
        }

        descr <- c("\\subsection{K-means Clustering}\n",
                "K-means clustering is a nonhierarchical clustering technique.",
                "It begins by creating k random clusters (k is supplied by user).",
                "The program then calculates the mean of each cluster.",
                "If an observation is closer to the centroid of another cluster",
                "then the observation is made a member of that cluster. This process is",
                "repeated until none of the observations are reassigned to a different cluster.",
                "\n\n",
                "K-means analysis is performed using the \\texttt{kmeans} function in the",
                "package \\texttt{stat}.",
                paste("Figure", fig.count<<-fig.count+1,"shows clustering the results."),
                paste("Table", table.count<<-table.count+1,"shows the members in each cluster from K-means analysis.\n"));

        cat(descr, file=rnwFile, append=TRUE);

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$kmeans,"}", sep=""),
                    "\\caption{K-means cluster analysis. The x-axes are variable indices and y-axes",
                    "are relative intensities. The blue lines represent median intensities of corresponding clusters}",
                    "\\end{center}",
                    paste("\\label{",imgSet$kmeans,"}", sep=""),
                    "\\end{figure}"
                    );
                    
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\n\n", file=rnwFile, append=TRUE, sep="\n");
        cmdhist2<-c("<<echo=false, results=tex>>=",
                    "GetAllKMClusterMembers()",
                    "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE);
}

# random forests
CreateRFdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$rf)){
                return();
        }

        descr <- c("\\subsection{Random Forest (RF) }\n",
                "Random Forest is a supervised learning algorithm suitable for high dimensional data analysis.",
                "It uses an ensemble of classification trees, each of which is grown by random feature",
                "selection from a bootstrap sample at each branch. Class prediction is based on the",
                "majority vote of the ensemble. RF also provides other useful information such as OOB",
                "(out-of-bag) error, variable importance measure, and outlier measures. During tree construction, about",
                "one-third of the instances are left out of the bootstrap sample. This OOB data",
                "is then used as test sample to obtain an unbiased estimate of the classification",
                "error (OOB error). Variable importance is evaluated by measuring the increase of the",
                "OOB error when it is permuted. The outlier measures are based on the proximities during tree construction.",
                "\n\n",
                "RF analysis is performed using the \\texttt{randomForest} package\\footnote{Andy Liaw and",
                "Matthew Wiener. \\textit{Classification and Regression by randomForest}, 2002, R News}.",
                paste("Table", table.count<<-table.count+1,"shows the confusion matrix of random forest."),
                paste("Figure", fig.count<<-fig.count+1,"shows the cumulative error rates of random forest analysis for given parameters.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the important features ranked by random forest.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the outlier measures of all samples for the given parameters.\n"));

        cat(descr, file=rnwFile, append=TRUE);

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$rf.cls,"}", sep=""),
                    "\\caption{Cumulative error rates by Random Forest classification. The overall error rate is shown
                        as the black line; the red and green lines represent the error rates for each class.}",
                    "\\end{center}",
                    paste("\\label{",imgSet$rf.cls,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c("<<echo=false, results=tex>>=",
                    "GetRFConf.Table()",
                    "@",
                    paste("The OOB error is ", GetRFOOB())
                );
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c( "\n\n",
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$rf.imp,"}", sep=""),
                    "\\caption{Significant features identified by Random Forest. The features are ranked by the mean
                             decrease in classification accuracy when they are permuted.}",
                    "\\end{center}",
                    paste("\\label{",imgSet$rf.imp,"}", sep=""),
                    "\\end{figure}"
                    );
                    
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c( "\n\n",
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$rf.outlier,"}", sep=""),
                    "\\caption{Potential outliers identified by Random Forest. Only the top five are labeled.}",
                    "\\end{center}",
                    paste("\\label{",imgSet$rf.outlier,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE);
}

# R-SVM
CreateSVMdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$svm)){
                return();
        }

        descr <- c("\\subsection{Support Vector Machine (SVM)}\n",
                "SVM aims to find a nonlinear decision function in the input space by mapping the data into a",
                "higher dimensional feature space and separating it there by means of a maximum margin hyperplane.",
                "The SVM-based recursive feature selection and classification is performed using the \\texttt{R-SVM}",
                "script\\footnote{http:\\slash\\slash www.hsph.harvard.edu\\slash bioinfocore\\slash RSVMhome\\slash R-SVM.html}.",
                "The process is performed recursively using decreasing series of feature subsets (\\texttt{ladder})",
                "so that different classification models can be calculated. Feature importance is evaluated based on",
                "its frequencies being selected in the best classifier identified by recursive classification and cross-validation.",
                "Please note, R-SVM is very computationally intensive. Only the top 50 features (ranked by their p values from t-tests)",
                "will be evaluated.",
                "\n\n",
                "In total,", length(analSet$svm$ladder), "models (levels) were created using",paste(analSet$svm$ladder, collapse=', '), "selected feature subsets.",
                paste("Figure", fig.count<<-fig.count+1, "shows the SVM classification performance using recursive feature selection."),
                paste("Figure", fig.count<<-fig.count+1, "shows the signicant features used by the best classifiers.\n"));

        cat(descr, file=rnwFile, append=TRUE);

        cat("\n\n", file=rnwFile, append=TRUE);
        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$svm.class,"}", sep=""),
                    "\\caption{Recursive classification with SVM. The red circle indicates the best classifier.}",
                    "\\end{center}",
                    paste("\\label{",imgSet$svm.class,"}", sep=""),
                    "\\end{figure}",
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$svm,"}", sep=""),
                    "\\caption{Significant features identified by R-SVM. Features are ranked by their frequencies of being selected in the classifer.}",
                    "\\end{center}",
                    paste("\\label{",imgSet$svm,"}", sep=""),
                    "\\end{figure}"
                    ); 
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE);
}


CreateStatFooter<-function(){

    descr <- c("\\section{Other Features}\n",
                "Please be advised that some more recent functions (Orthogonal PLS-DA, Power analysis, Biomarker analysis, Joint pathway analysis)",
                "with their corresponding results are not included in this report. For compound concentration data, users can perform",
                "metabolite set enrichment analysis and metabolic pathway analysis.",
                "\n\n");
    cat(descr, file=rnwFile, append=TRUE);

    cat("\\vspace{5 mm}\n", file=rnwFile, append=TRUE);
    cat("--------------------------------", file=rnwFile, append=TRUE);
    cat("\n\n", file=rnwFile, append=TRUE);
    end <- c("The report was generated on \\Sexpr{date()} with \\Sexpr{print(version$version.string)}.\n");
    cat(end, file=rnwFile, append=TRUE);
    cat("\\end{document}\n\n",file=rnwFile, append=TRUE);
}

###############################################
## Metabolite enrichment analysis report
##########################################

# write .Rnw file template
CreateEnrichRnwReport<-function(usrName){

    CreateHeader(usrName);
    CreateEnrichIntr();

    CreateEnrichOverview();
    CreateEnrichInputDoc();
    CreateEnrichProcessDoc();
    CreateEnrichAnalDoc();

    if(analSet$type == "msetora"){
        CreateEnrichORAdoc();
    }else if(analSet$type == "msetssp"){
        CreateEnrichSSPdoc();
        CreateEnrichORAdoc();
    }else{
        CreateEnrichQEAdoc();
    }

    CreateFooter();
}

CreateEnrichIntr<-function(){
        descr <- c("\\section{Background}\n",
                    "MESA or Metabolite Set Enrichment Analysis is a way to identify biologically meaningful",
                    "patterns that are significantly enriched in quantitative metabolomic data. In conventional",
                    "approaches, metabolites are evaluated individually for their significance under conditions",
                    "of study. Those compounds that have passed certain sigificance level are then combined to",
                    "see if any meaningful patterns can be discerned. In contrast, MSEA directly investigates if",
                    "a set of functionally related metabolites without the need to preselect compounds based on",
                    "some arbituary cut-off threshold. It has the potentail to identify subtle but consistent changes",
                    "among a group of related compounds, which may go undetected with the conventional approaches.",
                    "\n\n",
                    "Essentially, MSEA is a metabolomic version of the popular GSEA (Gene Set Enrichment Analysis)",
                    "software with its own collection of metabolite set libraries as well as an implementation of",
                    "user-friendly web-interfaces. GSEA is widely used in genomics data analysis and has proven to",
                    "be a powerful alternative to conventional approaches. For more information, please refer to",
                    "the original paper by Subramanian A, and a nice review paper by Nam D, Kim SY.\n"
                );
        cat(descr, file=rnwFile, append=TRUE);
}

CreateEnrichOverview<-function(){
        descr <- c("\\section{MSEA Overview}\n",
                   "Metabolite set enrichment analysis consists of four steps - data input, data processing,",
                   "data analysis, and results download. Different analysis procedures are performed based on",
                   "different input types. In addition, users can also browse and search the metabolite set",
                   "libraries as well as upload their self-defined metabolite sets for enrichment analysis.",
                   "Users can also perform metabolite name mapping between a variety of compound names, synonyms,",
                   "and major database identifiers."
                );
        cat(descr, file=rnwFile, append=TRUE);
}

# create data input doc
CreateEnrichInputDoc<-function(){

        descr <- c("\\section{Data Input}\n",
                    "There are three enrichment analysis algorithms offered by MSEA. Accordingly, three",
                    "different types of data inputs are required by these three approaches:\n"
                    );

        cat(descr, file=rnwFile, append=TRUE);
        descr <- c(
                "\\begin{itemize}",
                "\\item{A list of important compound names - entered as a one column data",
                 "(\\textit{Over Representation Analysis (ORA)});}",
                "\\item{A single measured biofluid (urine, blood, CSF) sample- entered as",
                "tab separated two-column data with the first column for compound name,",
                "and the second for concentration values",
                "(\\textit{Single Sample Profiling (SSP)});}",
                "\\item{A compound concentration table - entered as a comma separated (.csv) file",
                "with the each sample per row and each metabolite concentration per column.",
                "The first column is sample names and the second column for sample phenotype labels",
                "(\\textit{Quantitative Enrichment Analysis (QEA)})}",
                "\\end{itemize}",
                "\n\n");

        cat(descr, file=rnwFile, append=TRUE, sep="\n");

        if(analSet$type == "msetora"){
                descr <- c("You selected Over Representation Analysis (ORA) which requires a list of compound",
                    "names as input. \n\n"
                    );
        }else if(analSet$type == "msetssp"){
                descr <- c("You selected Single Sample Profiling (SSP) which requires a two-column data from",
                    "measurement of a single biofluid sample. Currently, only data from blood, urine and CSF",
                    "can be analysed as restricted by the availability of reference concentration data. \n\n"
                    );
        }else{
                descr <- c("You selected Quantitative Enrichment Analysis (QEA) which requires a concentration table.",
                    "This is the most common data format generated from quantitative metabolomcs studies.",
                    "The phenotype label can be can be discrete (binary or multi-class) or continuous. \n\n"
                    );
        }

        cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

CreateEnrichProcessDoc<-function(){
        descr <- c("\\section{Data Process}\n",
                   "The first step is to standardize the compound labels. It is an essential step since the compound",
                   "labels will be subsequently compared with compounds contained in the metabolite set library.",
                   "MSEA has a built-in tool to convert between compound common names, synonyms, identifiers used in",
                   "HMDB ID, PubChem, ChEBI, BiGG, METLIN, KEGG, or Reactome.",
                   "\\textbf{Table 1} shows the conversion results. Note: \\textit{1} indicates exact match, \\textit{2}",
                   "indicates approximate match, and \\textit{0} indicates no match. A text file contain the result can be",
                   "found the downloaded file \\textit{name\\_map.csv}\n\n"
                    );
       cat(descr, file=rnwFile, append=TRUE, sep="\n");

       descr<-c("<<echo=false, results=tex>>=",
                    "GetMapTable()",
                    "@",
                    "\\clearpage\n\n"
                    );
       cat(descr, file=rnwFile, append=TRUE, sep="\n");

       descr <- c("The second step is to check concentration values. For SSP analysis,",
                  "the concentration must be measured in \\textit{umol} for blood and CSF samples.",
                  "The urinary concentrations must be first converted to \\textit{umol/mmol\\_creatinine}",
                  "in order to compare with reported concentrations in literature. No missing or negative values",
                  "are allowed in SSP analysis. The concentration data for QEA analysis is more flexible.",
                  "Users can upload either the original concentration data or normalized data. Missing or negative values",
                  "are allowed (coded as \\textit{NA}) for QEA. Please note, MSEA does not perform data normalization.",
                  "If normalization is important, you should first normalize your data before upload.",
                  "You can use our companion website \\textbf{MetaboAnalyst} \\textit{www.metaboanalyst.ca} ",
                  "for a variety of data processing and normalization methods."
                 );
       cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

CreateEnrichAnalDoc<-function(){
        descr <- c("\\section{Selection of Metabolite Set Library}\n",
                   "Before proceeding to enrichment analysis, a metabolite set library has to be chosen.",
                   "There are seven built-in libraries offered by MSEA:\n"
                );
        cat(descr, file=rnwFile, append=TRUE, sep="\n");

        descr <- c(
                "\\begin{itemize}",
                "\\item{Metabolic pathway associated metabolite sets",
                 "(\\textit{currently contains 88 entries});}",
                "\\item{Disease associated metabolite sets (reported in blood)",
                "(\\textit{currently contains 416 entries});}",
                "\\item{Disease associated metabolite sets (reported in urine)",
                "(\\textit{currently contains 346 entries})}",
                "\\item{Disease associated metabolite sets (reported in CSF)",
                "(\\textit{currently contains 124 entries})}",
                "\\item{Metabolite sets associated with SNPs",
                "(\\textit{currently contains 4500 entries})}",
                "\\item{Predicted metabolite sets based on computational enzyme knockout model",
                "(\\textit{currently contains 912 entries})}",
                "\\item{Metabolite sets based on locations",
                "(\\textit{currently contains 57 entries})}",
                "\\end{itemize}",
                "\n\n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");

        descr <- c("In addition, MSEA also allows user-defined metabolite sets to be uploaded to perform",
                   "enrichment analysis on arbitrary groups of compounds which researchers want to test.",
                   "The metabolite set library is simply a two-column comma separated text file with the",
                   "first column for metabolite set names and the second column for its comound names (\\textbf{must use HMDB compound name})",
                   "separated by \"; \". Please note, the",
                   "built-in libraries are mainly from human studies. The functional grouping of metabolites",
                   "may not be valid. Therefore, for data from subjects other than human being, users are",
                   "suggested to upload their self-defined metabolite set libraries for enrichment analysis."
                );
        cat(descr, file=rnwFile, append=TRUE, sep="\n");
        cat("\\section{Enrichment Analysis}\n",file=rnwFile, append=TRUE, sep="\n");
}

CreateEnrichORAdoc<-function(){
        descr <- c(
                   "Over Representation Analysis (ORA) is performed when",
                   "a list of compound names is provided. The list of compound list can be obtained through",
                   "conventional feature selection methods, or from a clustering algorithm,",
                   "or from the compounds with abnormal concentrations detected in SSP,",
                   "to investigate if some biologically meaningful patterns can be identified.",
                   "\n\n",
                   "ORA was implemented using the \\textit{hypergeometric test} to evaluate whether a particular",
                   "metabolite set is represented more than expected by chance within the given compound list.",
                   "One-tailed p values are provided after adjusting for multiple testing. \\textbf{Figure 2} below",
                   "summarizes the result."
                );
        cat(descr, file=rnwFile, append=TRUE);
        fig <- c(  "\\begin{figure}[htp]",
                   "\\begin{center}",
                   paste("\\includegraphics[width=1.0\\textwidth]{",imgSet$ora,"}",sep=""),
                   "\\caption{Summary Plot for Over Representation Analysis (ORA)}",
                   "\\end{center}",
                   paste("\\label{",imgSet$ora,"}", sep=""),
                   "\\end{figure}",
                   "\\clearpage\n\n"
                    );
        cat(fig, file=rnwFile, append=TRUE, sep="\n");

        descr<-c("<<echo=false, results=tex>>=",
                    "GetORATable()",
                    "@",
                    "\\clearpage\n\n"
                    );
        cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

CreateEnrichSSPdoc<-function(){
          descr <- c(
                   "Single Sample Profiling (SSP) is designed to detect whether certain compound",
                   "concentrations measured from a particular sample are higher or lower compared",
                   "to their normal ranges reported in literature. Those selected",
                   "compound list are then subject to over representation analysis (ORA) to see",
                   "if certain biologically meaningful patterns can be identified. Please note,",
                   "those reference concentrations could be measured from heterogeneous analytical",
                   "technologies. It is advisable to compare only to those values that were",
                   "measured by similar technologies by referring to the original literature.",
                   "By default, if measured concentration values are higher or lower",
                   "than \\textbf{all} those reported concentration ranges will be selected for",
                   "enrichment analysis. Users can overwrite the default selections by manually",
                   "selecting or de-selecting them.",
                   "\n\n",
                   "\\textbf{Table 2} shows the comparison between the measured concentrations",
                   "and the reference concentrations. \\textit{L, M, H} means to the",
                   "measured concentration are \\textit{Lower, Within (Medium), Higher} compared",
                   "to the reference values. \\textit{0} means not selected for",
                   "subsequent enrichment analysis, while \\textit{1} means the corresponding",
                   "compound was selected."
                );
       cat(descr, file=rnwFile, append=TRUE, sep="\n")
       descr<-c("<<echo=false, results=tex>>=",
                    "GetSSPTable()",
                    "@",
                    "\\clearpage\n\n"
                );
       cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

CreateEnrichQEAdoc<-function(){
        descr <- c(
                   "Quantitative enrichment analysis (QEA) will be performed when the user uploads",
                   "a concentration table. The enrichment analysis is performed using package",
                   "\\textbf{globaltest} \\footnote{Jelle J. Goeman, Sara A. van de Geer, Floor de Kort",
                   "and Hans C. van Houwelingen.\\textit{A global test for groups of genes: testing",
                   "association with a clinical outcome}, Bioinformatics Vol. 20 no. 1 2004, pages 93-99}.",
                   "It uses a generalized linear model to estimate a",
                   "\\textit{Q-statistic} for each metabolite set, which describes the correlation",
                   "between compound concentration profiles, X, and clinical outcomes, Y. The \\textit{Q statistic}",
                   "for a metabolite set is the average of the Q statistics for each metabolite in the set.",
                   "\\textbf{Figure 2} below summarizes the result."
                );
        cat(descr, file=rnwFile, append=TRUE);
        fig <- c(  "\\begin{figure}[htp]",
                   "\\begin{center}",
                   paste("\\includegraphics[width=1.0\\textwidth]{",imgSet$qea, "}", sep=""),
                   "\\caption{Summary Plot for Quantitative Enrichment Analysis (QEA)}",
                   "\\end{center}",
                    paste("\\label{",imgSet$qea, "}", sep=""),
                   "\\end{figure}",
                   "\\clearpage\n\n"
                );
        cat(fig, file=rnwFile, append=TRUE, sep="\n");

        descr<-c("<<echo=false, results=tex>>=",
                   "GetQEATable()",
                   "@",
                   "\\clearpage\n\n"
                );
        cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

CreateFooter<-function(){

    end <- c("\\vspace{5 mm}\n--------------------------------\n\n",
             "The report was generated on \\Sexpr{date()} with \\Sexpr{print(version$version.string)}.\n",
             "\\end{document}\n\n"
            );
    cat(end, file=rnwFile, append=TRUE);
}

###############################################
## Metabolomic pathway analysis
##########################################

# write .Rnw file template
CreatePathRnwReport<-function(usrName){

    CreateHeader(usrName);
    CreatePathIntr();
    CreatePathInputDoc();
    CreatePathProcessDoc();
    CreatePathAnalDoc();
    CreatePathResultDoc();

    CreateFooter();
}

CreatePathIntr<-function(){
        descr <- c("\\section{Background}\n",
                    "The Pathway Analysis module combines results from powerful pathway",
                    "enrichment analysis with the pathway topology analysis to help researchers identify the most",
                    "relevant pathways involved in the conditions under study.",
                    "\n\n",
                    "There are many commercial pathway analysis software tools, such as Pathway Studio, MetaCore, or",
                    "Ingenuity Pathway Analysis (IPA), etc. Compared to them, the pathway analysis module was specially developed",
                    "for metabolomics studies. It uses the high-quality KEGG metabolic pathways as the backend knowledgebase.",
                    "It integrates many well-established (i.e. univariate analysis, over-representation analysis) methods,",
                    "as well as novel algorithms and concepts (i.e. Global Test, GlobalAncova, network topology analysis) into",
                    "pathway analysis. Another feature is a Google-Map style interactive visualization system to deliver",
                    "the analysis results in an intuitive manner. \n"
                );
        cat(descr, file=rnwFile, append=TRUE);
}

# create data input doc
CreatePathInputDoc<-function(){

        descr <- c("\\section{Data Input}\n",
                   "Pathway Analysis accepts either a list of compound labels (common names, HMDB IDs or KEGG IDs) with one compound per row,",
                   "or a compound concentration table with samples in rows and compounds in columns. The second column must be",
                   "phenotype labels (binary, multi-group, or continuous). The table is uploaded as comma separated values (.csv).",
                    "\n\n",
                    dataSet$data.msg,
                    "\n\n");

        cat(descr, file=rnwFile, append=TRUE);
}

CreatePathProcessDoc<-function(){
        descr <- c("\\section{Compound Name Matching}\n",
                   "The first step is to standardize the compound labels used in user uploaded data. This is a necessary step since",
                   "these compounds will be subsequently compared with compounds contained in the pathway library.",
                   "There are three outcomes from the step - exact match, approximate match (for common names only), and no match.",
                   "Users should click the textbf{View} button from the approximate matched results to manually select the correct one.",
                   "Compounds without match will be excluded from the subsequently pathway analysis.",
                   "\n\n",
                   "\\textbf{Table 1} shows the conversion results. Note: \\textit{1} indicates exact match, \\textit{2}",
                   "indicates approximate match, and \\textit{0} indicates no match. A text file contain the result can be",
                   "found the downloaded file \\textit{name\\_map.csv}\n\n"
                    );
       cat(descr, file=rnwFile, append=TRUE, sep="\n");

       descr<-c("<<echo=false, results=tex>>=",
                    "GetMapTable()",
                    "@",
                    "\\clearpage\n\n"
                    );
       cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

CreatePathAnalDoc<-function(){
        descr <- c("\\section{Pathway Analysis}\n",
                   "In this step, users are asked to select a pathway library, as well as specify the algorithms for pathway",
                   "enrichment analysis and pathway topology analysis. \n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");

        descr <- c("\\subsection{Pathway Library}\n",
                "There are 15 pathway libraries currently supported, with a total of 1173 pathways :\n",
                "\\begin{itemize}",
                "\\item{Homo sapiens (human) [80]}",
                "\\item{Mus musculus (mouse) [82]}",
                "\\item{Rattus norvegicus (rat) [81]}",
                "\\item{Bos taurus (cow) [81]}",
                "\\item{Danio rerio (zebrafish) [81]}",
                "\\item{Drosophila melanogaster (fruit fly) [79]}",
                "\\item{Caenorhabditis elegans (nematode) [78]}",
                "\\item{Saccharomyces cerevisiae (yeast) [65]}",
                "\\item{Oryza sativa japonica (Japanese rice) [83]}",
                "\\item{Arabidopsis thaliana (thale cress) [87]}",
                "\\item{Escherichia coli K-12 MG1655 [87]}",

                "\\item{Bacillus subtilis [80]}",
                "\\item{Pseudomonas putida KT2440 [89]}",
                "\\item{Staphylococcus aureus N315 (MRSA/VSSA)[73]}",
                "\\item{Thermotoga maritima [57]}",
                "\\end{itemize}",
                "\n\n",
                dataSet$lib.msg,
                "\n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");

        if(analSet$type == "pathora"){
            descr <- c("\\subsection{Over Representation Analysis}\n",
                "Over-representation analysis is to test if a particular group of compounds",
                "is represented more than expected by chance within the user uploaded compound",
                "list. In the context of pathway analysis, we are testing if compounds involved",
                "in a particular pathway is enriched compared by random hits. MetPA offers two",
                "most commonly used methods for over-representation analysis: \n",
                "\\begin{itemize}",
                "\\item{Fishers'Exact test}",
                "\\item{Hypergeometric Test}",
                "\\end{itemize}",
                "\\textit{Please note, MetPA uses one-tailed Fisher's exact test which will give essentially",
                "the same result as the result calculated by the hypergeometric test.}",
                "\n\n",
                analSet$rich.msg,
                "\n\n");
        }else{
            descr <- c("\\subsection{Pathway Enrichment Analysis}\n",
                "Pathway enrichment analysis usually refers to quantitative enrichment analysis directly",
                "using the compound concentration values, as compared to compound lists used by over-representation",
                "analysis. As a result, it is more sensitive and has the potentail to identify",
                "\\textbf{subtle but consistent} changes among compounds involved in the same biological pathway.",
                "\n\n",
                "Many procedures have been developed in the last decade for quantitative enrichment analysis, the most famous",
                "being the Gene Set Enrichment Analysis. Many new and improved methods have been implemented ever since.",
                "The enrichment analysis is based on GlobalTest and GlobalAncova. Both methods support enrichment analysis with",
                "binary, multi-group, as well as continuous phenotypes. The p values can be approximated based on the asymptotic",
                "distribution without using permutations which is computationally very intensive and is not suitable for web applications.",
                "Please note, when sample sizes are small, the approximated p values may be slightly less accurate compared to",
                "p values obtained by permutation-based method (for details, please refer to the paper by Goeman, J.J. et al.",
                "\\footnote{Jelle J. Goeman and Peter Buhlmann. \\textit{Analyzing gene expression data in terms of gene",
                "sets: methodological issues}, Bioinformatics 2007 23(8):980-987}",
                "and by Hummel, M. et al.",
                "\\footnote{Manuela Hummel, Reinhard Meister and Ulrich Mansmann. \\textit{GlobalANCOVA: exploration and assessment of gene group",
                "effects}, Bioinformatics 2008 24(1):78-85})",
                "However, since our focus is to identify the most relevant pathways within the pathways in the library,",
                "we are more interested in the rank of the pathway, not its absolote p-value. Therefore, this disadvantage may be tolerated.",
                "\n\n",
                analSet$rich.msg,
                "\n\n"
                );
        }
        cat(descr, file=rnwFile, append=TRUE, sep="\n");

        descr <- c("\\subsection{Pathway Topology Analysis}\n",
                 "The structure of biological pathways represent our knowledge about the complex relationships among molecules",
                 "within a cell or a living organism. However, most pathway analysis algorithms fail to take the structural information",
                 "into consideration when estimating which pathways are significantly changed under conditions of study",
                 "It is well-known that changes in more important positions of a network will trigger a more severe",
                 "impact on the pathway than changes occured in marginal or relatively isolated positions.",
                 "\n\n",
                 "The pathway toplogy analysis uses two well-established node centrality measures to estimate node importance - \\textbf{degree centrality}",
                 "and \\textbf{betweenness centrality}. Degree centrality is defined as the number of links occured upon a node.",
                 "For directed graph, there are two types of degree: in-degree for links come from other nodes, and out-degree",
                 "for links initiated from the current node. Metabolic networks are directed graph. Here we only consider the",
                 "out-degree for node importance measure. It is assumed that nodes in upstream will have regulatory roles for",
                 "the downstream nodes, not vice versa. The betweenness centrality measures number of shortest paths going",
                 "through the node. Since metabolic network is directed, we use relative betweenness centrality for metabolite",
                 "importance measure. The degree centrality measures focus more on local connectivities, while the betweenness",
                 "centrality measures focus more on global network topology. For more detailed discussions on various graph-based",
                 "methods for analysing biological networks, please refer to the article by Tero Aittokallio, T. et al.",
                 "\\footnote{Tero Aittokallio and Benno Schwikowski. \\textit{Graph-based methods for analysing networks in cell biology},",
                 "Briefings in Bioinformatics 2006 7(3):243-255}",
                 "\n\n",
                 "\\textit{Please note, for comparison among different pathways, the node importance values calculated from centrality measures",
                 "are further normalized by the sum of the importance of the pathway. Therefore,  the total/maximum importance of each pathway",
                 "is 1; the importance measure of each metabolite node is actually the percentage w.r.t the total pathway importance,",
                 "and the pathway impact value is the cumulative percentage from the matched metabolite nodes.}",
                 "\n\n",
                 analSet$topo.msg,
                 "\n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

CreatePathResultDoc<-function(){
        descr <- c("\\section{Pathway Analysis Result}\n",
                   "The results from pathway analysis are presented graphically as well as in a detailed table.",
                   "\n\n",
                   "A Google-map style interactive visualization system was implemented to facilitate data exploration.",
                   "The graphical output contains three levels of view: \\textbf{metabolome view}, \\textbf{pathway view},",
                   "and \\textbf{compound view}. Only the metabolome view is shown below.",
                   "Pathway views and compound views are generated dynamically based on your interactions with the",
                   "visualization system. They are available in your downloaded files. \n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");
        fig <- c(  "\\begin{figure}[htp]",
                   "\\begin{center}",
                   paste("\\includegraphics[width=1.0\\textwidth]{",imgSet$path.overview,"}",sep=""),
                   "\\caption{Summary of Pathway Analysis}",
                   "\\end{center}",
                   paste("\\label{",imgSet$path.overview,"}", sep=""),
                   "\\end{figure}",
                   "\\clearpage\n\n"
                    );
        cat(fig, file=rnwFile, append=TRUE, sep="\n");

        descr <- c(
                   "The table below shows the detailed results from the pathway analysis. Since",
                   "we are testing many pathways at the same time, the statistical p values from",
                   "enrichment analysis are further adjusted for multiple testings. In particular, ",
                   "the \\textbf{Total} is the total number of compounds in the pathway;",
                   "the \\textbf{Hits} is the actually matched number from the user uploaded data;",
                   "the \\textbf{Raw p} is the original p value calculated from the enrichment analysis;",
                   "the \\textbf{Holm p} is the p value adjusted by Holm-Bonferroni method;",
                   "the \\textbf{FDR p} is the p value adjusted using False Discovery Rate;",
                   "the \\textbf{Impact} is the pathway impact value calculated from pathway topology analysis.",
                   "\n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");
        if(analSet$type == "pathora"){
            descr<-c("<<echo=false, results=tex>>=",
                    "GetORATable()",
                    "@",
                    "\\clearpage\n\n"
                    );
        }else{
          descr<-c("<<echo=false, results=tex>>=",
                    "GetQEATable()",
                    "@",
                    "\\clearpage\n\n"
                    );
        }
        cat(descr, file=rnwFile, append=TRUE, sep="\n");

}


# write .Rnw file template
CreateTimeSeriesRnwReport<-function(usrName){

    CreateHeader(usrName);
    CreateTimeSeriesIOdoc();
    CreateStatNORMdoc();

    InitTimeSeriesAnal();
    if(exists("analSet")){
        CreateiPCAdoc();
        CreateHeatmap2doc();
        CreateAOV2doc();
        CreateASCAdoc();
        CreateMBdoc();
    }else{
        CreateTimeSeriesAnalNullMsg();
    }
    CreateFooter();
}


# read and process the raw data
CreateTimeSeriesIOdoc<-function(){
        descr <- c("\\section{Data Upload and Integrity Checking}\n");
        cat(descr, file=rnwFile, append=TRUE);

        descr <- c("\\subsection{Upload your data}\n",
                "For two-factor and time-series data, MetaboAnalyst accepts data uploaded in comma separated values (.csv) format.",
                "Samples can be in rows or columns. The two factor labels must follow immediately",
                "after the sample names. For time-series data, the time points group must be named",
                "as \\textbf{Time}. In addition, the samples collected from the same subject at different time points",
                "should be consecutive and ordered by the time points. Users need to specify the data types when uploading their data",
                "in order for MetaboAnalyst to select the correct algorithm to process them.\n",
                paste("Table", table.count<<-table.count+1,"summarizes the result of the data checking steps.\n")
                );
        cat(descr, file=rnwFile, append=TRUE);

        # error checking
        if(is.null(dataSet$orig) | is.null(dataSet$proc) | is.null(dataSet$facA) | is.null(dataSet$facB)){
                errorMsg<- c(descr, "Error occured during reading the raw data ....",
                    "Failed to proceed. Please check if the data format you uploaded is correct.",
                    "Please visit our FAQs, Data Formats, and TroubleShooting pages for more information!\n");
                cat(errorMsg, file=rnwFile, append=TRUE);
                return();
        }

        cat("\n\n", file=rnwFile, append=TRUE);

        descr <- paste(dataSet$read.msg, collapse="\n");
        cat(paste("\\texttt{", PrepareLatex(descr), "}"), file=rnwFile, append=TRUE);
        cat("\n\n", file=rnwFile, append=TRUE);

        cmdhist2<-c("<<echo=false, results=tex>>=",
                    "CreateSummaryTable()",
                    "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");

        # the next step is sanity check
        descr<-c("\\subsection{Data Integrity Check}\n",
                "Before data analysis, a data integrity check is performed to make sure that all the necessary",
                "information has been collected.",
                "Compound concentration or peak intensity values should all be non-negative numbers.",
                "By default, all missing values, zeros and negative values will be replaced by the half of the minimum positive value",
                "found within the data (detection limits).\n\n");
        cat(descr, file=rnwFile, append=TRUE);

        msgLen <- length(dataSet$check.msg);
        descr <- paste(dataSet$check.msg[1:(msgLen-2)], collapse="\n");
        cat(paste("\\texttt{", PrepareLatex(descr), "}"), file=rnwFile, append=TRUE);
        cat("\n\n", file=rnwFile, append=TRUE);
}

InitTimeSeriesAnal<-function(){
        descr <- c("\\section{Statistical and Machine Learning Data Analysis}",
                   "For two-factor and time-series data, MetaboAnalyst offers several carefully selected methods for general two-factor and time-series",
                   "data analysis. They include:\n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");

        descr2 <- c(
                "\\begin{itemize}",

                "\\item{Data overview: }",
                "\\begin{itemize}",
                "\\item{Interactive Principal Component Analysis (iPCA) }",
                "\\item{Two-way Heatmap clustering and visualization}",
                "\\end{itemize}",

                "\\item{Univariate method: }",
                "\\begin{itemize}",
                "\\item{Two-way between/within-subjects ANOVA}",
                "\\end{itemize}",

                "\\item{Multivariate approaches}",
                "\\begin{itemize}",
                "\\item{ANOVA-Simultaneous Component Analysis (ASCA)}",
                "\\item{Multivariate Empirical Bayes Analysis (MEBA)}",
                "\\end{itemize}",

                "\\end{itemize}",

                "\\texttt{Please note: MEBA is only applicable to time-series data analyais.}",
                "\\clearpage"
            );
        cat(descr2, file=rnwFile, append=TRUE, sep="\n");
}

CreateTimeSeriesAnalNullMsg<-function(){
        descr <- c("No analysis was performed on your data.\n");
        cat(descr, file=rnwFile, append=TRUE, sep="\n");
}

# create iPCA doc
CreateiPCAdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$ipca)){
                return();
        }

        descr <- c("\\subsection{Interactive Principal Component Analysis (iPCA)}\n",
                "PCA is an unsupervised method aiming to find the directions that best",
                "explain the variance in a data set (X) without referring to class labels (Y).",
                "The data are summarized into top three components or PCs that explain most of the",
                "variations in the data. The result is displayed in an interactive 3D",
                "visualization system. The system supports pointing-and-clicking, rotating",
                "zooming(hold down SHIFT key and drag vertically). Clicking any of the displayed",
                "data points will show the corresponding sample summarized by the top 20 most",
                "different variables that devidate from the data center.",
                "\n\n");

       cat(descr, file=rnwFile, append=TRUE);
       cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");
}


# two-way heatmap
CreateHeatmap2doc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$htmap2)){
                return();
        }

        descr <- c("\\subsection{Two-way Heatmap Visualization}\n",
                "The heatmap provides direct visualization of all the data points in the forms",
                "of colors squares. The color spectrum intuitively indicates the higher or lower values",
                "Users can choose different clustering algorithms or distance measures to cluster the",
                "variables. The samples are ordered by the two factors with default the first factor",
                "used for primary ordering. Users can choose to switch the order.",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the clustering result in the form of a heatmap.\n"));

        cat(descr, file=rnwFile, append=TRUE);

        if(!is.null(analSet$htmap2)){
            cmdhist<-c(
                "\\begin{figure}[htp]",
                "\\begin{center}",
                paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$htmap2,"}", sep=""),
                paste("\\caption{Clustering result shown as heatmap (",
                      "distance measure using ", "\\texttt{", analSet$htmap2$dist.par, "}, and clustering algorithm using ", "\\texttt{", analSet$htmap2$clust.par, "}).}", sep=""),
                "\\end{center}",
                paste("\\label{",imgSet$htmap2,"}", sep=""),
                "\\end{figure}"
                );
            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        }
        cat("\\clearpage", file=rnwFile, append=TRUE);
}

# create Univartests doc
CreateAOV2doc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$aov2)){
                return();
        }

        if(isEmptyMatrix(analSet$aov2$sig.mat)){
            aov2.tab<-NULL;
        }else{
            aov2.tab<-paste("Table", table.count<<-table.count+1,"shows the details of these features;");
        }


        descr <- c("\\subsection{Univariate Analysis}\n",
                "Univariate analysis methods are the most common methods used for exploratory data analysis. ",
                "For two-factor data, the basic approach is two-way ANOVA",
                "There are two options - between-subjects ANOVA and within-subjects ANOVA. When samples are all from",
                "indepent subjects (i.e. general two-way ANOVA), the between-subjects option should be selected.",
                "The time series data contains samples measured from the same subjects from different time points.",
                "The within-subjects ANOVA should be used.",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the important features identified by ANOVA analysis."),
                aov2.tab,
                "\n");

        cat(descr, file=rnwFile, append=TRUE);

        cmdhist<-c(
                  "\\begin{figure}[htp]",
                  "\\begin{center}",
                  paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$anova2,"}", sep=""),
                  "\\caption{Important features selected by two-way ANOVA}",
                  "\\end{center}",
                  paste("\\label{",imgSet$anova2,"}", sep=""),
                  "\\end{figure}"
              );
        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        cat("\n\n", file=rnwFile, append=TRUE, sep="\n");

        cmdhist2<-c("<<echo=false, results=tex>>=",
                   "GetSigTable.Aov2()",
                   "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE, sep="\n");
}



# random forests
CreateASCAdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$asca)){
                return();
        }

        if(isEmptyMatrix(analSet$asca$sig.list[["Model.a"]])){
            asca.tab1<-NULL;
        }else{
            asca.tab1<-paste("Table", table.count<<-table.count+1,paste("shows features well-modelled by ", dataSet$facA.lbl, ". ", sep=""));
        }

        if(isEmptyMatrix(analSet$asca$sig.list[["Model.b"]])){
            asca.tab2<-NULL;
        }else{
            asca.tab2<-paste("Table", table.count<<-table.count+1,paste("shows features well-modelled by ", dataSet$facB.lbl, ". ", sep=""));
        }

        if(isEmptyMatrix(analSet$asca$sig.list[["Model.ab"]])){
            asca.tab3<-NULL;
        }else{
            asca.tab3<-paste("Table", table.count<<-table.count+1, "shows features well-modelled by Interaction model. ");
        }

        descr <- c("\\subsection{ANOVA - Simultaneous Component Analysis (ASCA)}\n",
                "ASCA is a multivariate extension of univariate ANOVA approach. It is",
                "designed to identify the major patterns associated with each factor.",
                "This implementation supports ASCA model for two factors with one interaction",
                "effect. The algorithm first partitions the overall data variance (X) into",
                "individual variances induced by each factor (A and B), as well as by the",
                "interactions (AB). The formula is shown below with (E) indicates the residual",
                "Errors: \n\n",
                "\\textbf{X = A + B + AB + E}",
                "\n\n",
                "The SCA part applies PCA to A, B, AB to summarize major variations in each partition.",
                "Users then detect the major pattern by visualizing the PCA scores plot.",
                "MetaboAnalyst also provides model validation to test the significance of the effects associated",
                "with main effects. It is based on the Manly's unrestricted permutation of observation",
                "then calculate the permuted variation associated with each factor",
                "Finally, the permuted values are compared with the original variations",
                "The significant variables are identified based on the leverage and the",
                "Squared Prediction Errors (SPE) assciated with each variables",
                "Variables with low SPE and higher leverage are modelled well after the major patterns.",
                "\n\n",
                paste("Figure", fig.count<<-fig.count+1,"shows the scree plots for each effect model.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the major patterns associated with factor A.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the major patterns associated with factor B.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the major patterns associated with interaction.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the results of model validations through permutations.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the important features associated with factor A.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the important features associated with factor B.\n"),
                paste("Figure", fig.count<<-fig.count+1,"shows the features that are important in the interaction.\n"),
                "\n\n",
                asca.tab1,
                asca.tab2,
                asca.tab3,
                "The other details are available as .csv documents in your downloaded zip file.");

        cat(descr, file=rnwFile, append=TRUE);
        cat("\\clearpage", file=rnwFile, append=TRUE);

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$asca.scree,"}", sep=""),
                    "\\caption{Scree plots for each sub model}",
                    "\\end{center}",
                    paste("\\label{",imgSet$asca.scree,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$asca.modelA,"}", sep=""),
                    paste("\\caption{Major patterns associated with", dataSet$facA.lbl, "}"),
                    "\\end{center}",
                    paste("\\label{",imgSet$asca.modelA,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$asca.modelB,"}", sep=""),
                    paste("\\caption{Major patterns associated with", dataSet$facB.lbl, "}"),
                    "\\end{center}",
                    paste("\\label{",imgSet$asca.modelB,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$asca.modelAB,"}", sep=""),
                    "\\caption{Major patterns associated with the Interaction between the two factors}",
                    "\\end{center}",
                    paste("\\label{",imgSet$asca.modelAB,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$asca.perm,"}", sep=""),
                    "\\caption{Model validation through permutations}",
                    "\\end{center}",
                    paste("\\label{",imgSet$asca.perm,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$asca.impA,"}", sep=""),
                    paste("\\caption{Important variables associated with", dataSet$facA.lbl, "}"),
                    "\\end{center}",
                    paste("\\label{",imgSet$asca.impA,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$asca.impB,"}", sep=""),
                    paste("\\caption{Important variables associated with", dataSet$facB.lbl, "}"),
                    "\\end{center}",
                    paste("\\label{",imgSet$asca.impB,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cmdhist<-c(
                    "\\begin{figure}[htp]",
                    "\\begin{center}",
                    paste("\\includegraphics[width=1.0\\textwidth]{", imgSet$asca.impAB,"}", sep=""),
                    "\\caption{Variables important in interaction between the two factors}",
                    "\\end{center}",
                    paste("\\label{",imgSet$asca.impAB,"}", sep=""),
                    "\\end{figure}"
                    );

        cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");

        cat("\\clearpage\n\n", file=rnwFile, append=TRUE);

        if(!is.null(asca.tab1)){
            cmdhist<-c("<<echo=false, results=tex>>=",
                   "GetSigTable.ASCA(\"Model.a\")",
                   "@");
            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        }

        if(!is.null(asca.tab2)){
            cmdhist<-c("<<echo=false, results=tex>>=",
                   "GetSigTable.ASCA(\"Model.b\")",
                   "@");
            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        }

        if(!is.null(asca.tab3)){
            cmdhist<-c("<<echo=false, results=tex>>=",
                   "GetSigTable.ASCA(\"Model.ab\")",
                   "@");
            cat(cmdhist, file=rnwFile, append=TRUE, sep="\n");
        }
        cat("\\clearpage", file=rnwFile, append=TRUE);
}

# MEBA
CreateMBdoc<-function(){

        # need to check if this process is executed
        if(is.null(analSet$MB)){
                return();
        }

        descr <- c("\\subsection{Multivariate Empirical Bayes Approach - MEBA}\n",
                "The approach is designed to compare the time-course profiles under different conditions.",
                "The result is a list of variables that are ranked by their difference in temporal profiles",
                "across different biological conditions. The Hotelling-T2 is used to rank the variables with",
                "different temporal profiles between two biological conditions under study; And the",
                "MB-statistics is used for more than two biological conditions. Higher statistical value",
                "indicates the time-course profiles are more different across the biological conditions under study.",
                "\n");

        cat(descr, file=rnwFile, append=TRUE);

        cmdhist2<-c("<<echo=false, results=tex>>=",
                   "GetSigTable.MB()",
                   "@");
        cat(cmdhist2, file=rnwFile, append=TRUE, sep="\n");
        cat("\\clearpage", file=rnwFile, append=TRUE);
}



